File: python/src/gitea_toolkit_mcp/operations/issues.py
"""Issue 域操作（移植自 TS 版 `src/core/operations/issues.ts`）。

Gitea 里 Issue 与 PR 共用底层表：``/repos/issues/search`` 用 ``type`` 参数区分两者，
本模块只处理 Issue（PR 见 ``pulls.py``）。
"""

from __future__ import annotations

from typing import Any, Literal
from urllib.parse import quote

from ..client import GiteaClient, ListResult

#: 状态取值；``all`` 表示不过滤。
#:
#: 用 ``Literal`` 而不是 ``str``：生成的 JSON Schema 里会带 enum，
#: 客户端据此渲染下拉框，也避免模型传入服务端不认的取值。
IssueState = Literal["open", "closed", "all"]

#: 条目类型（Issue 与 PR 共用接口时的区分参数）。
IssueSearchType = Literal["issues", "pulls"]


def _enc(value: str) -> str:
    """URL 路径片段编码。"""
    return quote(str(value), safe="")


def _flag(value: bool | None) -> str | None:
    """把布尔开关转成 Gitea 需要的查询值。

    必须显式写 ``"true"``：``urlencode`` 会把 Python 的 ``True`` 序列化成字面量
    ``True``，服务端不认，会当成「未提供」而静默忽略过滤条件。
    """
    return "true" if value else None


class IssueOperations:
    """Issue、评论与标签。"""

    def __init__(self, client: GiteaClient) -> None:
        self._client = client

    async def list(
        self,
        *,
        owner: str | None = None,
        repo: str | None = None,
        state: IssueState = "open",
        item_type: IssueSearchType = "issues",
        search: str | None = None,
        labels: str | None = None,
        milestones: str | None = None,
        assigned_to_me: bool | None = None,
        created_by_me: bool | None = None,
        mentioned_me: bool | None = None,
        review_requested_me: bool | None = None,
        reviewed_by_me: bool | None = None,
        sort: str | None = None,
        page: int = 1,
        limit: int = 50,
    ) -> ListResult:
        """列出 Issue。

        ``owner`` 与 ``repo`` 同时给出时走单仓库接口；否则走全局检索接口
        （「分配给我的 Issue」「提及我的 Issue」这类跨仓库场景只能用后者，
        过滤参数也只有后者支持）。
        """
        # 走 request_paged：服务端单页上限是 max_response_items（默认 50），
        # 想要更多必须真的翻页 —— 只调大 limit 会被**静默截断**。
        common: dict[str, Any] = {
            "state": state,
            "type": item_type,
            "q": search,
            "labels": labels,
            "milestones": milestones,
            "sort": sort,
        }
        extract = lambda data: data if isinstance(data, list) else []  # noqa: E731

        if owner and repo:
            return await self._client.request_paged(
                "GET",
                f"/repos/{_enc(owner)}/{_enc(repo)}/issues",
                limit=limit,
                start_page=page,
                query=common,
                extract=extract,
            )

        return await self._client.request_paged(
            "GET",
            "/repos/issues/search",
            limit=limit,
            start_page=page,
            query={
                **common,
                "assigned": _flag(assigned_to_me),
                "created": _flag(created_by_me),
                "mentioned": _flag(mentioned_me),
                "review_requested": _flag(review_requested_me),
                "reviewed": _flag(reviewed_by_me),
                "owner": owner,
            },
            extract=extract,
        )

    async def get(self, owner: str, repo: str, index: int) -> dict[str, Any]:
        """取单个 Issue 详情。"""
        return await self._client.request(
            "GET", f"/repos/{_enc(owner)}/{_enc(repo)}/issues/{index}"
        )

    async def create(
        self,
        owner: str,
        repo: str,
        *,
        title: str,
        body: str | None = None,
        assignees: list[str] | None = None,
        labels: list[int] | None = None,
        milestone: int | None = None,
        due_date: str | None = None,
        closed: bool | None = None,
        ref: str | None = None,
    ) -> dict[str, Any]:
        """创建 Issue。"""
        payload: dict[str, Any] = {"title": title}
        if body is not None:
            payload["body"] = body
        if assignees:
            payload["assignees"] = assignees
        if labels:
            payload["labels"] = labels
        if milestone is not None:
            payload["milestone"] = milestone
        if due_date:
            payload["due_date"] = due_date
        if closed is not None:
            payload["closed"] = closed
        if ref:
            payload["ref"] = ref
        return await self._client.request(
            "POST", f"/repos/{_enc(owner)}/{_enc(repo)}/issues", body=payload
        )

    async def update(
        self,
        owner: str,
        repo: str,
        index: int,
        *,
        title: str | None = None,
        body: str | None = None,
        state: Literal["open", "closed"] | None = None,
        assignees: list[str] | None = None,
        labels: list[int] | None = None,
        milestone: int | None = None,
        due_date: str | None = None,
        unset_due_date: bool = False,
    ) -> dict[str, Any]:
        """更新 Issue（只提交**显式给出**的字段）。

        ``labels`` 传空列表表示「清空标签」—— 所以调用方要用 ``None`` 区分「不改」与「清空」。
        """
        # ⚠️ 标签**不能**塞进 PATCH 的 body：实例 swagger 里 ``EditIssueOption`` 根本没有
        #    ``labels`` 字段（10 个字段里没有它），传了会被服务端**静默忽略** ——
        #    调用方以为改了标签、实际没改（TS 版就有这个问题）。
        #    标签有专门的替换端点，见下方。
        #    顺序刻意放在 PATCH **之前**：这样 PATCH 返回的实体里带的就是更新后的标签，
        #    不用再多发一次 GET 去取最新状态。
        if labels is not None:
            await self.replace_labels(owner, repo, index, labels)

        payload: dict[str, Any] = {}
        if title is not None:
            payload["title"] = title
        if body is not None:
            payload["body"] = body
        if state is not None:
            payload["state"] = state
        if assignees is not None:
            payload["assignees"] = assignees
        if milestone is not None:
            payload["milestone"] = milestone
        if due_date is not None:
            payload["due_date"] = due_date
        if unset_due_date:
            payload["unset_due_date"] = True
        return await self._client.request(
            "PATCH", f"/repos/{_enc(owner)}/{_enc(repo)}/issues/{index}", body=payload
        )

    async def replace_labels(
        self, owner: str, repo: str, index: int, labels: list[int]
    ) -> None:
        """整体替换 Issue 的标签（空列表 = 清空）。

        单独暴露成方法，是为了让「标签走专用端点」这件事在调用处看得见 ——
        它**不是** PATCH issue 的一部分（见 ``update`` 里的说明）。
        """
        await self._client.request(
            "PUT",
            f"/repos/{_enc(owner)}/{_enc(repo)}/issues/{index}/labels",
            body={"labels": labels},
        )

    async def list_comments(
        self, owner: str, repo: str, index: int, *, page: int = 1, limit: int = 50
    ) -> ListResult:
        """列出 Issue 的评论（翻页收集，见 ``request_paged`` 的说明）。"""
        return await self._client.request_paged(
            "GET",
            f"/repos/{_enc(owner)}/{_enc(repo)}/issues/{index}/comments",
            limit=limit,
            start_page=page,
            extract=lambda data: data if isinstance(data, list) else [],
        )

    async def create_comment(
        self, owner: str, repo: str, index: int, body: str
    ) -> dict[str, Any]:
        """在 Issue（或 PR 的对话区）发表评论。"""
        return await self._client.request(
            "POST",
            f"/repos/{_enc(owner)}/{_enc(repo)}/issues/{index}/comments",
            body={"body": body},
        )

    async def list_labels(self, owner: str, repo: str) -> list[Any]:
        """列出仓库标签（供创建 / 更新 Issue 时把标签名解析成 ID）。

        用 ``request_paged`` 而非单次 ``limit=200`` 请求：服务端单页上限默认 50，
        单次请求在标签多的仓库上会**静默少一半**，导致部分标签名解析不出 ID。
        """
        result = await self._client.request_paged(
            "GET",
            f"/repos/{_enc(owner)}/{_enc(repo)}/labels",
            limit=200,
            extract=lambda data: data if isinstance(data, list) else [],
        )
        return list(result.items)


__all__ = ["IssueOperations", "IssueSearchType", "IssueState"]

File: src/core/operations/issues.ts
/**
 * Issue 相关的 Gitea API 操作（Gitea 中 Issue 与 PR 共用底层表，此处只处理 Issue）。
 */
import type { GiteaClient } from '../giteaClient';
import type {
  GiteaComment,
  GiteaIssue,
  GiteaIssueSearchType,
  GiteaIssueState,
  GiteaLabel,
  GiteaListResult,
} from '../types';

/** Issue 列表查询参数。 */
export interface ListIssuesOptions {
  /** 限定仓库所属者；与 `repo` 同时提供时走单仓库接口。 */
  owner?: string;
  /** 限定仓库名；与 `owner` 同时提供时走单仓库接口。 */
  repo?: string;
  /** 状态筛选，默认 `open`。 */
  state?: GiteaIssueState;
  /** 条目类型：issue 或 pull。默认 issues。 */
  type?: GiteaIssueSearchType;
  /** 关键词。 */
  search?: string;
  /** 标签名，逗号分隔。 */
  labels?: string;
  /** 里程碑名，逗号分隔。 */
  milestones?: string;
  /** 只看分配给我的。 */
  assignedToMe?: boolean;
  /** 只看我创建的。 */
  createdByMe?: boolean;
  /** 只看提及我的。 */
  mentionedMe?: boolean;
  /** 只看请求我评审的（PR）。 */
  reviewRequestedMe?: boolean;
  /** 只看我评审过的（PR）。 */
  reviewedByMe?: boolean;
  /** 排序字段：`created` / `updated` / `comments` / `due_date`（仅单仓库接口支持）。 */
  sort?: string;
  page?: number;
  limit?: number;
}

/** 创建 Issue 参数。 */
export interface CreateIssueOptions {
  /** 标题（必填）。 */
  title: string;
  /** 正文（Markdown）。 */
  body?: string;
  /** 指派用户。 */
  assignees?: string[];
  /** 标签 ID 列表。 */
  labels?: number[];
  /** 里程碑 ID。 */
  milestone?: number;
  /** 截止日期，ISO 8601。 */
  dueDate?: string;
  /** 直接以关闭状态创建。 */
  closed?: boolean;
  /** 关联的分支名。 */
  ref?: string;
}

/** 更新 Issue 参数。 */
export interface UpdateIssueOptions {
  title?: string;
  body?: string;
  /** `open` 或 `closed`。 */
  state?: 'open' | 'closed';
  assignees?: string[];
  labels?: number[];
  milestone?: number;
  dueDate?: string;
  /** 清除截止日期。 */
  unsetDueDate?: boolean;
}

/** Issue 操作集合。 */
export class IssueOperations {
  constructor(private readonly client: GiteaClient) {}

  /**
   * 列出 Issue（可按仓库或全局检索）。
   * @param options 查询参数
   * @returns Issue 列表与分页信息
   */
  public async list(options: ListIssuesOptions = {}): Promise<GiteaListResult<GiteaIssue>> {
    // 走 requestPaged：服务端单页上限为 max_response_items（默认 50），
    // 想要更多必须真的翻页，只调大 limit 会被静默截断。
    const state = options.state ?? 'open';
    const limit = options.limit ?? 50;
    const startPage = options.page ?? 1;

    if (options.owner && options.repo) {
      return this.client.requestPaged<GiteaIssue>(
        'GET',
        `/repos/${enc(options.owner)}/${enc(options.repo)}/issues`,
        {
          limit,
          startPage,
          extract: (data) => (data as GiteaIssue[]) ?? [],
          query: {
            state,
            type: options.type,
            q: options.search,
            labels: options.labels,
            milestones: options.milestones,
            sort: options.sort,
          },
        },
      );
    }

    return this.client.requestPaged<GiteaIssue>('GET', '/repos/issues/search', {
      limit,
      startPage,
      extract: (data) => (data as GiteaIssue[]) ?? [],
      query: {
        state,
        type: options.type,
        q: options.search,
        labels: options.labels,
        milestones: options.milestones,
        assigned: options.assignedToMe ? 'true' : undefined,
        created: options.createdByMe ? 'true' : undefined,
        mentioned: options.mentionedMe ? 'true' : undefined,
        review_requested: options.reviewRequestedMe ? 'true' : undefined,
        reviewed: options.reviewedByMe ? 'true' : undefined,
        owner: options.owner,
      },
    });
  }

  /**
   * 获取单个 Issue。
   * @param owner 所属者
   * @param repo 仓库名
   * @param index Issue 序号（`number`）
   * @returns Issue 详情
   */
  public get(owner: string, repo: string, index: number): Promise<GiteaIssue> {
    return this.client.request<GiteaIssue>('GET', `/repos/${enc(owner)}/${enc(repo)}/issues/${index}`);
  }

  /**
   * 创建 Issue。
   * @param owner 所属者
   * @param repo 仓库名
   * @param options 创建参数
   * @returns 新建的 Issue
   */
  public create(owner: string, repo: string, options: CreateIssueOptions): Promise<GiteaIssue> {
    const body: Record<string, unknown> = { title: options.title };
    if (options.body !== undefined) body.body = options.body;
    if (options.assignees?.length) body.assignees = options.assignees;
    if (options.labels?.length) body.labels = options.labels;
    if (options.milestone !== undefined) body.milestone = options.milestone;
    if (options.dueDate) body.due_date = options.dueDate;
    if (options.closed !== undefined) body.closed = options.closed;
    if (options.ref) body.ref = options.ref;
    return this.client.request<GiteaIssue>('POST', `/repos/${enc(owner)}/${enc(repo)}/issues`, { body });
  }

  /**
   * 更新 Issue（标题 / 正文 / 状态 / 指派 / 标签 / 里程碑）。
   * @param owner 所属者
   * @param repo 仓库名
   * @param index Issue 序号
   * @param options 更新参数
   * @returns 更新后的 Issue
   */
  public update(
    owner: string,
    repo: string,
    index: number,
    options: UpdateIssueOptions,
  ): Promise<GiteaIssue> {
    const body: Record<string, unknown> = {};
    if (options.title !== undefined) body.title = options.title;
    if (options.body !== undefined) body.body = options.body;
    if (options.state !== undefined) body.state = options.state;
    if (options.assignees !== undefined) body.assignees = options.assignees;
    if (options.labels !== undefined) body.labels = options.labels;
    if (options.milestone !== undefined) body.milestone = options.milestone;
    if (options.dueDate !== undefined) body.due_date = options.dueDate;
    if (options.unsetDueDate) body.unset_due_date = true;
    return this.client.request<GiteaIssue>('PATCH', `/repos/${enc(owner)}/${enc(repo)}/issues/${index}`, { body });
  }

  /**
   * 列出 Issue 的评论。
   * @param owner 所属者
   * @param repo 仓库名
   * @param index Issue 序号
   * @param page 页码
   * @param limit 每页数量
   * @returns 评论列表
   */
  public async listComments(
    owner: string,
    repo: string,
    index: number,
    page = 1,
    limit = 50,
  ): Promise<GiteaListResult<GiteaComment>> {
    const result = await this.client.requestWithMeta<GiteaComment[]>(
      'GET',
      `/repos/${enc(owner)}/${enc(repo)}/issues/${index}/comments`,
      { query: { page, limit } },
    );
    return { items: result.data ?? [], pageInfo: result.pageInfo };
  }

  /**
   * 在 Issue 下发表评论。
   * @param owner 所属者
   * @param repo 仓库名
   * @param index Issue 序号
   * @param body 评论正文（Markdown）
   * @returns 新建的评论
   */
  public createComment(owner: string, repo: string, index: number, body: string): Promise<GiteaComment> {
    return this.client.request<GiteaComment>(
      'POST',
      `/repos/${enc(owner)}/${enc(repo)}/issues/${index}/comments`,
      { body: { body } },
    );
  }

  /**
   * 列出仓库标签（供创建 Issue 时选择）。
   * @param owner 所属者
   * @param repo 仓库名
   * @returns 标签列表
   */
  public async listLabels(owner: string, repo: string): Promise<GiteaLabel[]> {
    const result = await this.client.requestWithMeta<GiteaLabel[]>(
      'GET',
      `/repos/${enc(owner)}/${enc(repo)}/labels`,
      { query: { limit: 200 } },
    );
    return result.data ?? [];
  }

}

/** URL 路径片段编码。 */
function enc(value: string): string {
  return encodeURIComponent(value);
}
